docs/src/main/asciidoc/picocli.adoc

```diff
@@ -0,0 +1,238 @@
+class GoodByeCommand implements Runnable {
+
+    @Override
+    public void run() {
+        System.out.println("Goodbye World!");
+    }
+}
+----
+
+== Customizing Picocli CommandLine instance
+
+You can customize CommandLine classes used by the `picocli` extension by producing your own bean instance:
+
+[source,java]
+----
+package com.acme.picocli;
+
+import io.quarkus.picocli.runtime.PicocliCommandLineFactory;
+import io.quarkus.picocli.runtime.annotations.TopCommand;
+import picocli.CommandLine;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+
+@TopCommand
+@CommandLine.Command
+public class EntryCommand implements Runnable {
+    @CommandLine.Spec
+    CommandLine.Model.CommandSpec spec;
+
+    @Override
+    public void run() {
+        System.out.println("My name is: " + spec.name());
+    }
+}
+
+@ApplicationScoped
+class CustomConfiguration {
+
+    @Produces
+    CommandLine customCommandLine(PicocliCommandLineFactory factory) { // <1>
+        return factory.create().setCommandName("CustomizedName");
+    }
+}
+----
+<1> `PicocliCommandLineFactory` will create an instance of CommandLine with `TopCommand` and `CommandLine.IFactory` injected.
+
+== Different entry command for each profile
+
+It is possible to create different entry command for each profile, using `@IfBuildProfile`:
+
+[source,java]
+----
+@ApplicationScoped
+public class Config {
+
+    @Produces
+    @TopCommand
+    @IfBuildProfile("dev")
+    public Object devCommand() {
+        return DevCommand.class; // <1>
+    }
+
+    @Produces
+    @TopCommand
+    @IfBuildProfile("prod")
+    public Object prodCommand() {
+        return new ProdCommand("Configured by me!");
+    }
+
+}
+----
+<1> You can return instance of `java.lang.Class` here. In such case `CommandLine` will try to instantiate this class using `CommandLine.IFactory`.
+
+== Providing own QuarkusMain
+
+You can also provide your own application entry point annotated with `QuarkusMain` (as described in link:command-mode-reference[Command Mode reference guide]).
+
+[source,java]
+----
+package com.acme.picocli;
+
+import io.quarkus.runtime.QuarkusApplication;
+import io.quarkus.runtime.annotations.QuarkusMain;
+import picocli.CommandLine;
+
+import javax.inject.Inject;
+
+@QuarkusMain
+@CommandLine.Command(name = "demo", mixinStandardHelpOptions = true)
+public class ExampleApp implements Runnable, QuarkusApplication {
+    @Inject
+    CommandLine.IFactory factory; // <1>
+
+    @Override
+    public void run() {
+        // business logic
+    }
+
+    @Override
+    public int run(String... args) throws Exception {
+        return new CommandLine(this, factory).execute(args);
+    }
+}
+----
+<1> Quarkus-compatible `CommandLine.IFactory` bean created by `picocli` extension.
+
+== Native mode support
+
+GraalVM Native images should work out-of-box, but in case of any issues it is still possible to use annotation processor from `picocli` project.
```

Comment:
I am a little hesitant about mentioning this TBH... Nowhere in any other extension are we telling users to add some additional process to get their GraalVM native image working. 
We should strive to make everything work in native image and treat things that don't as bugs to be fixed here, instead of relying on an annotation processor.

Reply:
I agree that if something doesn't work in native images created with buildsteps then it is a bug and must be fixed. But, if I understood @remkop correctly, `picocli-codegen` may have some improvements in future to make native images smaller by tweaking reflection configuration. With Quarkus, users won't be able to benefit from those improvements until they will be ported to buildsteps, which is more like feature and not bugfix. 

As for now we probably can remove this property and decided what to do when improvements will be added to `picocli-codegen`.

Reply:
Well, in my view if we want to tweak the native image stuff for picocli, then the best way to do that would be to have the proper hooks in picocli so that Quarkus could do it's own optimizations (much in the same way that is already the case for other extensions)

Reply:
I'm little lost here. Do you mean hooks in `picocli` extension or `picocli` library itself? Current implementation of buildsteps is quite aggressive in adding reflection to native image. For example all methods are added from @Command annotated classes, even if they are not called by reflection in runtime. And @remkop also mentioned possible changes in picocli library itself.

Reply:
I mean that picocli could (might already ?) Have hooks that the Quarkus extension could use to optimize things (create bytecode for implementing interfaces for example).
This is of course something for future consideration

Reply:
> As for now we probably can remove this property and decided what to do when improvements will be added to picocli-codegen.

😭  I had concerns with build steps replacing `picocli-codegen` because this potentially limits how picocli can evolve as a library (without breaking picocli in Quarkus). My concerns were addressed by adding this property. 

The purpose of this property was to provide a fallback solution for any interim periods of time when picocli has evolved in an incompatible way, and the Quarkus picocli extension has not been updated yet. 

The existence of this fallback mechanism is important to me.  Please reinstate it.

> I am a little hesitant about mentioning this TBH... 

Now, how this fallback mechanism is _documented_ is a different matter.
The intention is certainly not to advertise this as canonical or normal usage.

I am fine with very minimal documentation or even omitting this from the documentation.